### PR TITLE
Enable RTL as mission item

### PR DESCRIPTION
**Pull request by tumbili**
Tested in HIL, seems to work fine.

### CODE REVIEW

File: src/modules/navigator/mission_block.cpp [hunk 1]
```diff
@@ -211,6 +212,13 @@ MissionBlock::mission_item_to_position_setpoint(const struct mission_item_s *ite
...
 
+	case NAV_CMD_RETURN_TO_LAUNCH:
+		sp->type = position_setpoint_s::SETPOINT_TYPE_RTL;
+		sp->lat = _navigator->get_home_position()->lat;
+		sp->lon = _navigator->get_home_position()->lon;
+		sp->alt = _navigator->get_global_position()->alt;
+		break;
+
 	default:
 		sp->type = position_setpoint_s::SETPOINT_TYPE_POSITION;
```
Comment (LorenzMeier) [inline src/modules/navigator/mission_block.cpp]: The default case here still needs to be fixed. We just covered this one corner case, but we didn't fix the root bug, which is to *NOT* have a default statement like this one, but to skip the (unsupported) item and continue with the next supported mission item (or abort / end the mission - that's an architecture question).

**Comment by tumbili** [comment]
@LorenzMeier HIL tested but will test this today outdoors.

**Comment by tumbili** [comment]
@LorenzMeier Tested this yesterday. Still need to fix a small bug.

**Comment by LorenzMeier** [comment]
Do we know the fix already?

**Comment by LorenzMeier** [comment]
Moved to next release. Thanks for fixing the corner cases!

File: src/modules/navigator/mission_feasibility_checker.cpp [hunk 2]
```diff
@@ -74,11 +75,16 @@ bool MissionFeasibilityChecker::checkMissionFeasible(bool isRotarywing, dm_item_
 		_mavlink_fd = open(MAVLINK_LOG_DEVICE, 0);
 	}
 
+	// check if all mission item commands are supported
+	failed |= !checkMissionItemValidity(dm_current, nMissionItems);
```
Comment (bansiesta) [inline src/modules/navigator/mission_feasibility_checker.cpp]: bitwise or used for bool, I would say this needs to be:
```
failed = failed || !checkMissionItemValidity(dm_current, nMissionItems);
```

File: src/modules/navigator/mission_feasibility_checker.cpp [hunk 3]
```diff
@@ -74,11 +75,16 @@ bool MissionFeasibilityChecker::checkMissionFeasible(bool isRotarywing, dm_item_
...
+
 
 	if (isRotarywing)
-		return checkMissionFeasibleRotarywing(dm_current, nMissionItems, geofence, home_alt);
+		failed |= !checkMissionFeasibleRotarywing(dm_current, nMissionItems, geofence, home_alt);
```
Comment (bansiesta) [inline src/modules/navigator/mission_feasibility_checker.cpp]: more bitwise

File: src/modules/navigator/mission_feasibility_checker.cpp [hunk 4]
```diff
@@ -74,11 +75,16 @@ bool MissionFeasibilityChecker::checkMissionFeasible(bool isRotarywing, dm_item_
...
 	else
-		return checkMissionFeasibleFixedwing(dm_current, nMissionItems, geofence, home_alt);
+		failed |= !checkMissionFeasibleFixedwing(dm_current, nMissionItems, geofence, home_alt);
```
Comment (bansiesta) [inline src/modules/navigator/mission_feasibility_checker.cpp]: more bitwise

File: src/modules/navigator/mission_feasibility_checker.cpp [hunk 5]
```diff
@@ -163,6 +169,39 @@ bool MissionFeasibilityChecker::checkHomePositionAltitude(dm_item_t dm_current,
...
+			missionitem.nav_cmd != NAV_CMD_WAYPOINT &&
+			missionitem.nav_cmd != NAV_CMD_LOITER_UNLIMITED &&
+			missionitem.nav_cmd != NAV_CMD_LOITER_TURN_COUNT &&
+			missionitem.nav_cmd != NAV_CMD_LOITER_TIME_LIMIT &&
+			missionitem.nav_cmd != NAV_CMD_RETURN_TO_LAUNCH &&
+			missionitem.nav_cmd != NAV_CMD_LAND &&
+			missionitem.nav_cmd != NAV_CMD_TAKEOFF &&
+			missionitem.nav_cmd != NAV_CMD_ROI &&
+			missionitem.nav_cmd != NAV_CMD_PATHPLANNING &&
+			missionitem.nav_cmd != NAV_CMD_DO_JUMP) {
```
Comment (bansiesta) [inline src/modules/navigator/mission_feasibility_checker.cpp]: Good idea to check for unsupported things! 

What about using a switch here and then putting the return false in the default?

File: src/modules/navigator/mission_feasibility_checker.cpp [hunk 6]
```diff
@@ -163,6 +169,39 @@ bool MissionFeasibilityChecker::checkHomePositionAltitude(dm_item_t dm_current,
 	return true;
 }
 
+bool MissionFeasibilityChecker::checkMissionItemValidity(dm_item_t dm_current, size_t nMissionItems) {
```
Comment (bansiesta) [inline src/modules/navigator/mission_feasibility_checker.cpp]: I would call it it checkMissionItemValid(). I think that's a bit more intuitive for what to expect from the return bool, and it's the same style as "feasible".

File: src/modules/navigator/mission.cpp [hunk 7]
```diff
@@ -228,12 +231,15 @@ Mission::update_offboard_mission()
 		 * however warnings are issued to the gcs via mavlink from inside the MissionFeasiblityChecker */
 		dm_item_t dm_current = DM_KEY_WAYPOINTS_OFFBOARD(_offboard_mission.dataman_id);
 
-		_missionFeasiblityChecker.checkMissionFeasible(_navigator->get_vstatus()->is_rotary_wing,
+		failed = !_missionFeasiblityChecker.checkMissionFeasible(_navigator->get_vstatus()->is_rotary_wing,
 				dm_current, (size_t) _offboard_mission.count, _navigator->get_geofence(),
 				_navigator->get_home_position()->alt);
 
 	} else {
 		warnx("offboard mission update failed");
```
Comment (bansiesta) [inline src/modules/navigator/mission.cpp]: shouldn't you also set failed = true in here?

Comment (tumbili) [inline src/modules/navigator/mission.cpp]: Thanks for the feedback, been busy with the release but will come back to this soon!

**Comment by LorenzMeier** [comment]
@tumbili @julianoes Should we take bets if it takes less than a year to finish this for good? 8).

**Comment by LorenzMeier** [comment]
@tumbili Time flies! Almost one year for this one!

**Pull request closed without merging**